### ISSUE TITLE
Branded types

### ISSUE THREAD

**Issue by steida** [fp-ts/schema#98]
Hi, and thank you for your awesome work! I have a question related to branded types.
I'm not sure which approach I should choose:

1) RefinedConstructor

```ts
type Int = number & B.Brand<"Int">
const Int = B.refined<Int>(
  (n) => Number.isInteger(n),
  (n) => B.error(`Expected ${n} to be an integer`)
)
const IntSchema = pipe(S.number, _.brand(Int))
```

2) Schema with filters and nominal type

```ts
type Int = number & Brand<"Int">;
const Int = pipe(S.number, S.between(1, 4), brand(nominal<Int>()));
```

From the average Schema user perspective, it seems 2) is less verbose.
And it could be even less verbose with some brand helper.

```ts
// some brand helper
const Int = pipe(S.number, S.between(1, 4), brand<'Int'>());
type Int = S.Infer<typeof Int>
```

WDYT?

**Comment by IMax153** [comment]
I think that all of the options posted above are valid depending on the use case. 

However, I think I would probably choose the first option - reason being you can get all the benefits of the `Schema` instance while also retaining access to the direct constructors when needed:

```ts
Int.of(...)
Int.option(...)
const decodeInt = Schema.decode(IntSchema)
```

**Comment by steida** [comment]
@IMax153 Why is retaining access to the direct constructors needed when Schema provides the same API? I'm just curious, thank you.

**Comment by jessekelly881** [comment]
I don't think this interface `brand<'Int'>()` is supported currently but it could be added. I think both interfaces are fine. For library code I would prefer the first option. The code for the first option is actually even more verbose because you need to define annotations for the brand as well. So, it would look like 

```
const IntSchema = pipe(S.number, _.brand(Int, {
    jsonSchema: { type: "integer" },
}))
```

**Comment by gcanti** [comment]
I think the `brand` combinator of `/schema` is useful when a user has already defined their own brand for other reasons and wants to reuse it in a schema.

Otherwise if the user is starting from scratch defining a new schema, a cast could be sufficient.

```ts
import * as B from "@effect/data/Brand";
import { pipe } from "@fp-ts/core/Function";
import * as S from "@fp-ts/schema";

type Int = number & B.Brand<"Int">

const Int = B.refined<Int>(
  (n) => Number.isInteger(n),
  (n) => B.error(`Expected ${n} to be an integer`)
)

//
// reuse
//

// $ExpectType Schema<number & Brand<"Int">>
pipe(S.number, S.brand(Int))

//
// cast
//

// $ExpectType Schema<Int>
pipe(S.number, S.int(), S.as<Int>())
```

**Comment by gcanti** [comment]
@jessekelly881 as suggested by @steida we could add a `brand` API (similar to zod's `brand`)

```ts
// Schema.ts

declare const brand: <B extends string, A>(
  brand: B,
  options?: AnnotationOptions<A>
) => (self: Schema<A>) => Schema<A & Brand<B>> 
```

and accumulate brands in a specific annotation (compilers like TypeScript may want to read this info):

```ts
// const UserId: S.Schema<number & Brand<"UserId">>
const UserId = pipe(
  S.number,
  S.int(),
  S.brand("UserId", {
    description: "A user id"
  })
)
expect(UserId.ast.annotations).toEqual({
  [A.BrandId]: ["UserId"],
  [A.DescriptionId]: "A user id",
  [A.JSONSchemaId]: { type: "integer" }
})
```

https://github.com/fp-ts/schema/blob/3b362ac931a64585a058325a8926b1be75416635/test/Schema.ts#L24-L35

**Comment by jessekelly881** [comment]
Overloading the brand function to support both brands and strings complicates it quite a bit. Do you think we should make this a separate function?

**Comment by gcanti** [comment]
> we could add a brand API

released (v0.2.1) so we can have early feedback.

Closing
